core/cat/headers.py

```diff
@@ -18,8 +24,32 @@
 
 api_key_header = APIKeyHeader(name="access_token", auto_error=False)
 
+def ws_auth(
+    websocket: WebSocket,
+    ) -> None | str:
+    """Authenticate endpoint.
+
+    Check the provided key is available in API keys list.
+
+    Parameters
+    ----------
+    request : Request
+        HTTP request.
+
+    Returns
+    -------
+    api_key : str | None
+        Returns the valid key if set in the `.env`, otherwise return None.
+
+    Raises
+    ------
+    HTTPException
+        Error with status code `403` if the provided key is not valid.
 
-def check_api_key(request: Request, api_key: str = Security(api_key_header)) -> None | str:
+    """
+    return websocket.app.state.ccat.authorizator.is_ws_allowed(websocket)
```

Comment:
In this case you still need to check the master_key as in the http request, otherwise the admin panel will break if the dev implements a custom rule in is_allowed_ws

Reply:
I'm not so sure about this, I try to expand a little bit my consideration.
If the idea is to keep backward compatible the actual behavior, with this solution the websocket remains open.
So setting an API_KEY any endpoint will be protected and the websocket not.
What do you think?
I involve even @zAlweNy26 that knows how the admin panel works.

Reply:
i don't agree with keeping websocket chat open, this is a terrible security issue. If a dev wants to implement custom security autorizer for the websocket connection, it should not be public but the admin still needs to work.
To allow that we always have to somehow pass the api key from the admin panel.

In the default authorizer "BaseAuth" the check on is_ws_allowed returns true so by default the old admin should works until the dev implement an authorizer with custom is_ws_allowed logic.

To avoid having backward compatibility problems we need to extend the admin panel to somehow pass the api key in websocket connection.

Little recap:
with currrent websocket check implementation -> authorizator.is_ws_allowed(websocket)
using BaseAuth:
-> **old admin panel** will works because is_ws_allowed returns true (✅) (No backward compatibility)
-> **direct websocket** connection should works without api key because is_ws_allowed returns true (✅)
using CustomAuth with custom is_ws_allowed logic
-> **admin panel** should not work because the admin panel doesn't know the authorization logic (❌)
-> **direct websocket** connection should works implementing custom dev authorizer logic (✅)

implementing this websocket check -> authorizator.is_ws_master_key(websocket) or authorizator.is_ws_allowed(websocket)
using BaseAuth:
-> **old admin panel** should still works because is_ws_allowed returns true (✅) (No backward compatibility)
-> **direct websocket** connection should works both with and without api key (✅)
using CustomAuth with custom is_ws_allowed logic
-> **old admin panel** should not work (Backward compatibility -> **we need to pass api key**) (❌)
-> **direct websocket** connection should works implementing custom dev authorizer logic (✅)


what do you think? @giovannialbero1992 @zAlweNy26

Reply:
I think we are overcomplicating because two things happen:
- the Cat is designed as a machine to machine microservice
- the admin is not machine to machine

We currently understand what is going on but this could be really hard to get for newbies.
As a long term vision, I'd love the Cat to be used standalone, and *also* as a microservice.
Most of all, it has to be all around easy to use :)

I take it as my own fault, for insisting so much on both having no users, and having the admin included in core.
I take it also on my shoulders to merge this PR and directly adding the following (help needed only on the admin):

## Admin user

There will be an user system into core, and as a start it will only include the `admin` user.
Credentials for the admin will be changeable via two env variables:
`CCAT_ADMIN_USER` // defaults to `admin`
`CCAT_ADMIN_PASSWORD` // defaults to `admin`
This is a raw start for a full fledged user system, that can evolve without breaking the admin panel.

When admin panel opens, it can do as follows:
- search for a JWT in localStorage, if not
- shows a username / password modal
- sends credentials to the core and receives a temp JWT

JWT is used for:
- http requests via `Authorization` header
- ws communication via a "token" key included in each message (or only on connection via URL parameter)

## Machine 2 machine

As it is now, `API_KEY` will cover the http endpoints and websocket stays open.
We could easily add a `WS_API_KEY` to also secure websockets.

In M2M communication, as far as I know, there is no need for a JWT and keys can be used directly.
Nothing prevents to use the user/pass credential system also for M2M.

## Summary

- Browsers are authenticated via credentials and JWT
- M2M is authenticated both like a browser or simply via static keys.
- Auth can be customized at will as we already designed, and the admin panel is saved.

Custom auth stays as @giovannialbero1992 implemented it, with minor modifications.
I'm going all in, tell me if you see something I'm not seeing.